scripts/fuse/README.MD

```diff
@@ -0,0 +1,166 @@
+
+A side effect is that creating/deleting a single key can create/delete multiple "virtual" directories at once.
+Another is that copies created with `cp -r` will not have these vacuous intermediaries but instead empty keys.
+
+### Keys that act both as file and directory
+
+In Elektra, there are only keys and no distinction between directories and files are made. To enable a bidirectional mapping into a classical file system, the following mechanism is employed:
+
+If a key (`A`) exists, and has some child key (`B`), `A` is treated as a directory containing `B` and a special file `@elektra.value`. This file is used to access the value of `A`.
+
+Some side effects arise:
+E.g. recursively deleting directories with `rm -r` containing `@elektra.value` files raises (task uncritical) errors when attempting to delete a parent directory which was already deleted with the removal of an `@elektra.value` file directly below.
+
+### "Leaf"-Keys that act as as directories
+
+Keys at the bottom of the hierarchy are normally treaded as files. However in certain cases, like explicitly creating directories using `mkdir`, this would result in files beeing displayed, making it impossible to expand the hierarchy using standard tools.
+Therefore, when a directory is created, the special meta-key `fuse-directory` is set to override this behaviour.
+
+### Filesystem attributes
+
+The filesystem attributes are sourced from the file obtained by `kdb file`.
+For some keys, no such file is returned, or the returned file does not exist in actuality. In these cases, fake attributes are used to preserve usability.
+
+To increase compatibility, currently the timestamps from these files are not used, as this prompts software like `vim` to produce warnings/errors.
+
+## Meta-Keys
+
+Meta-keys are mapped bidirectionally to extended file attributes (xattrs).
+To enable interoperability with other tools, the `meta:/` prefix is optional when reading/writing accessing meta keys. The listing of attributes however will show them prefix-less.
+
+Furthermore, no binary values for these attributes are allowed. (As libelektra does not accept those).
+
+See `xattr(1)` on how to read/write/list these attributes.
+
+## Binary-Keys
+
+If a value is to be written that cannot be decoded using the systems default encoding, it is treated as binary, and the appropriate `meta:/binary` meta-key is set.
+
+## Known issues
+
+- Moving an ``@elektra.value``-file corrupts the filesystem/key database (See this [issue](https://issues.libelektra.org/3648)). As editors like `vim` may move a file during editing, such editors cannot reliably be used on these pseudo-files.
+- Only functional on UNIX-like systems
+- chmod/chown is not implemented and does not signal a "not supported error" to enable compatibility with common tools like `cp`.
+- Access related errorcodes not yet implemented (e.g. normal user tries to write to system:/) (defaulting to read only fs - error)
```

Comment:
I see reporting Elektra's errors to the user within the scope of the thesis.

Reply:
As far as I know FUSE does not support custom error messages, we only have the standard UNIX-error codes available.

Reply:
Most file systems simply log what they wanted to say if they cannot via errno, e.g. they report file corruption that way. KDB reporting errors is very similar as it is also "the layer below". You could simply use [syslog](https://docs.python.org/3/library/syslog.html).

Special files would be another option to report errors (like lost+found), but that would conflict with your stateless approach.

Reply:
Just to avoid any misunderstandings:

I am supposed to catch (unexpected, e.g. binary writes to meta-keys) Elektra exceptions and log them via syslog in my implementation?

Reply:
No, things like binary writes to meta-keys can be done via normal errno codes maybe with ENOTSUP.

The logger would be used when kdbOpen/kdbGet/kdbSet return errors.

Reply:
As of 83d67af36959a3181e9641ccd1cf299ce10961f5 there is configurable logging (via command line options:(stdout/syslog/none)).  Exceptions thrown by Elektra are also logged.